Fix NoEcho parameters get replaced with ****

danmactough opened the pull request:
Resolves #177

To summarize the problem, any parameters marked with the `"NoEcho": true` property get changed to the literal value `****` when we perform a stack update and accept the previous value for those parameters.

The general approach here is to use the `UsePreviousValue` property instead of providing the previous value as `ParameterValue` when we build our ChangeSet. Currently, I don't think there's a 100% foolproof way to know from the `newParameters` whether a user has accepted the previous value instead of providing a new value because all we know is that the value is `****`. This strikes me as too generic (meaning, too likely that a user may want the literal value `****`), so when we fetch the previous parameters from CloudFormation, I've replaced `****` with a very unlikely replacement string: `@@__CFN-CONFIG__NOECHO__PLACEHOLDER__@@`. Then, when we're building our ChangeSet, if we see that string, we know not to provide it as the `ParameterValue` in the ChangeSet and instead to specify the `"UsePreviousValue": true` property for that parameter.

In addition, when we write our configuration to S3, we now check each parameter to see if it has the `NoEcho` property and if so, we replace its value with `****` before writing it to S3. Previously, on initial stack creation, the plaintext value of NoEcho parameters would be in the configuration stored on S3 and then, depending on user behavior, potentially overwritten by `****`. Both writing secrets in plaintext and having variable behavior on update are undesirable, so the change here makes sure that we consistently mask NoEcho parameters. This means we can never use a saved configuration to reuse the value of NoEcho parameters, but (a) it wouldn't have worked reliably before (so IMO it's not a breaking change) and (b) it actually seems to me like a good idea to not allow those secrets to be reused that way.

Note that using the `@@__CFN-CONFIG__NOECHO__PLACEHOLDER__@@` placeholder is not strictly necessary. I can back that out which would simplify these changes a bit. Thoughts?

cc @rclark

rclark commented:
Oh there's a PR too :)

During prompting, couldn't we

a) understand the default value that we're about to provide to the user
b) accept the user input
c) if the user input === the default value, then `UsePreviousValue: true`

... with the caveat that we'd also have to know that the default _came from existing stack info_ instead of that it came from the template itself.

The interplay of code names to represent this case seems overly complex, at least at first read.

danmactough commented:
> The interplay of code names to represent this case seems overly complex, at least at first read.

Roger that. I'll back that out.

> During prompting, couldn't we
> 
> a) understand the default value that we're about to provide to the user
> b) accept the user input
> c) if the user input === the default value, then `UsePreviousValue: true`
> 
> ... with the caveat that we'd also have to know that the default _came from existing stack info_ instead of that it came from the template itself.

@rclark I'm not sure I understand exactly how you're thinking about this working, so check me if I'm off-the-mark, here. I agree that it would be preferable to switch stack updates to use `UsePreviousValue: true` in all cases when a user does not update a parameter.

At [the point in the code where we make need to make that user input <=> default value comparision](https://github.com/mapbox/cfn-config/blob/0cc06b5fadaf2c669874f07db04d4d62edfc538d/lib/actions.js#L330-L343), we don't currently have access to the `context` or old parameters. But if we change [this call to `actions.diff`](https://github.com/mapbox/cfn-config/blob/0cc06b5fadaf2c669874f07db04d4d62edfc538d/lib/commands.js#L504-L511), to also pass in `context.oldParams`, we can then pass that along to the `stackParameters` function. So, I think that would work.

rclark commented:
If we go all the way back to where the [`context.newParameters` is set](https://github.com/mapbox/cfn-config/blob/0cc06b5fadaf2c669874f07db04d4d62edfc538d/lib/commands.js#L369), it seems like we are in a position where we could do this analysis of the user's input. This context property looks like it ends up being the `parameters` object that becomes the `UpdateStack` API call.

If that flat key-value object was actually designed to be the full parameter input for the API call (along with `UsePreviousValue: true` where appropriate), would it pass though ok?

danmactough commented:
> If that flat key-value object was actually designed to be the full parameter input for the API call (along with UsePreviousValue: true where appropriate), would it pass though ok?

We do have all the context we need at that point. But there are a couple places where context.newParameters is expected to be that flat key-value structure that would also need to be updated:

- https://github.com/mapbox/cfn-config/blob/0cc06b5fadaf2c669874f07db04d4d62edfc538d/lib/commands.js#L353-L358
- https://github.com/mapbox/cfn-config/blob/0cc06b5fadaf2c669874f07db04d4d62edfc538d/lib/commands.js#L379
- https://github.com/mapbox/cfn-config/blob/0cc06b5fadaf2c669874f07db04d4d62edfc538d/lib/commands.js#L706-L714

That last one would be really adversely impacted by changing the structure of `context.newParameters` as it wants all the parameter values (not just changed ones) to write out to s3. We could deal with it, but it would be pretty ugly (like, building up that flat key-value parameters object before we pass it into `actions.saveConfiguration`).

If you feel strongly about performing that analysis at that point, maybe we should leave `context.newParameters` as is and when we build the full parameter input for the API call using `UsePreviousValue` where appropriate at that point, set it on a **new** context property -- `context.changesetParameters` maybe.

rclark commented:
> If you feel strongly about performing that analysis at that point, 

... I kind of do ... which granted isn't the strongest statement, but I would really prefer this.

> maybe we should leave context.newParameters as is and when we build the full parameter input for the API call using UsePreviousValue where appropriate at that point, set it on a new context property -- context.changesetParameters maybe.

This sounds like a decent approach.

danmactough commented:
@rclark Updated

- backed out the magic placeholder value via a5d6609
- implemented creating that ChangeSet parameters array via 75a41bb 3326d18
- added a new test in 94be3d5

rclark commented:
Ag sorry @danmactough this fell off the radar this week. I'll circle back soon.

danmactough commented:
![gong-mr-rogers](https://user-images.githubusercontent.com/357481/88863491-f9a58b00-d1d0-11ea-840d-c6d1b469c037.gif)

rclark commented on lib/commands.js:
```diff
@@ -341,23 +343,29 @@ var operations = module.exports.operations = {
+    var overrideParameters = {};
 
     if (context.overrides.parameters) {
       Object.keys(context.overrides.parameters).forEach(function(key){
         if (newTemplateParameters[key] || context.oldParameters[key])
-          context.oldParameters[key] = context.overrides.parameters[key];
+          overrideParameters[key] = context.overrides.parameters[key];
```
So the reason that we set the context object's old and new parameter values to the override value is so that parameter changes that are passed as overrides do not show up as part of the diff between old/new parameter sets. [This comparison ends up being done in the `confirmParameters` function](https://github.com/mapbox/cfn-config/blob/0cc06b5fadaf2c669874f07db04d4d62edfc538d/lib/commands.js#L379).

Could you keep this part that sets the override value in `context.oldParameters`, but collect all the TRUE old values in its own object so that you can pass the un-overriden content to `changesetParameters`?

danmactough replied:
> So the reason that we set the context object's old and new parameter values to the override value is so that parameter changes that are passed as overrides do not show up as part of the diff between old/new parameter sets. [This comparison ends up being done in the `confirmParameters` function](https://github.com/mapbox/cfn-config/blob/0cc06b5fadaf2c669874f07db04d4d62edfc538d/lib/commands.js#L379).

I think we've already accounted for this.

In the [if/else clause](https://github.com/mapbox/cfn-config/pull/178/files#diff-5e05cf0e6359f91b6a5859876734d7f0L352-R368) right below here, we either:

- When `(context.overrides.force || context.overrides.skipPromptParameters)` is `true`, we **do not** assign the `overrideParameters` to `context.oldParameters`. This is the source of your concern. However, if you look at [`confirmParameters`](https://github.com/mapbox/cfn-config/pull/178/files#diff-5e05cf0e6359f91b6a5859876734d7f0R383-R388) you'll notice that it continues early on the same condition -- when `(context.overrides.force || context.overrides.skipPromptParameters)` is `true` -- and doesn't not perform the diff.
- When `(context.overrides.force || context.overrides.skipPromptParameters)` is `false`, we **DO** assign the `overrideParameters` to `context.oldParameters`. Then in `confirmParameters`, `context.oldParameters` has the expected values when the diff is performed.

Does this address your concern @rclark?

rclark replied:
Yeah, it does. Thanks @danmactough.

danmactough commented:
> What you're fixing is arguably a bug, but it feels major because of the changes to the API calls being made to CloudFormation.

Totally agree. It also changes the configuration data that's saved off to S3. Wrong or not, people could have been depending on that behavior.
